src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs

```diff
@@ -155,11 +155,18 @@ public override async Task<OrchestrationHistory> GetHistoryEventsAsync(string in
                 .GetHistoryEntitiesResponseInfoAsync(instanceId, expectedExecutionId, null, cancellationToken)
                 .GetResultsAsync(cancellationToken: cancellationToken);
 
+            // The sentinel row should always be the last row
+            TableEntity sentinel = results.Entities.LastOrDefault(e => e.RowKey == SentinelRowKey);
+
             IList<HistoryEvent> historyEvents;
             string executionId;
-            TableEntity sentinel = null;
             TrackingStoreContext trackingStoreContext = new TrackingStoreContext();
-            if (results.Entities.Count > 0)
+
+            // If expectedExecutionId is provided but it does not match the sentinel executionId,
+            // it may belong to a previous generation. In that case, treat it as an unknown executionId
+            // and skip loading history.
+            if (results.Entities.Count > 0 && (expectedExecutionId == null ||
+                                               expectedExecutionId == sentinel?.GetString("ExecutionId")))
             {
                 // The most recent generation will always be in the first history event.
                 executionId = results.Entities[0].GetString("ExecutionId");
```

Comment:
Would it make sense to change this line as well to something like the following?

```suggestion
                executionId = sentinel?.GetString("ExecutionId") ?? results.Entities[0].GetString("ExecutionId");
```

I'm just wondering if we want to find a way to ensure we never attempt to load irrelevant table entities. The check above ensures that we don't load history for messages that have an old execution ID, but I wonder if there could be other situations where we're unintentionally loading irrelevant table entities. I think this likely won't happen because we won't have a sentinel that doesn't match the first entity we get back (see line 179) but it's something we could potentially consider as defense-in-depth.